### PR TITLE
swift 5.7

### PR DESCRIPTION
**Pull request by Bo98**
Created by `brew update-swift`. 🚀

---

Created with `brew bump-formula-pr`.

`resource` blocks may require updates.

### CODE REVIEW

**Comment by danielnachun** [comment]
@Bo98 it's failing on Linux because we need to pass `-fPIC` to `libCyaml` when it is being built:
```
FAILED: lib/libYams.so Sources/Yams/CMakeFiles/Yams.dir/Constructor.swift.o Sources/Yams/CMakeFiles/Yams.dir/Decoder.swift.o Sources/Yams/CMakeFiles/Yams.dir/Emitter.swift.o Sources/Yams/CMakeFiles/Yams.dir/Encoder.swift.o Sources/Yams/CMakeFiles/Yams.dir/Mark.swift.o Sources/Yams/CMakeFiles/Yams.dir/Node.Mapping.swift.o Sources/Yams/CMakeFiles/Yams.dir/Node.Scalar.swift.o Sources/Yams/CMakeFiles/Yams.dir/Node.Sequence.swift.o Sources/Yams/CMakeFiles/Yams.dir/Node.swift.o Sources/Yams/CMakeFiles/Yams.dir/Parser.swift.o Sources/Yams/CMakeFiles/Yams.dir/Representer.swift.o Sources/Yams/CMakeFiles/Yams.dir/Resolver.swift.o Sources/Yams/CMakeFiles/Yams.dir/String+Yams.swift.o Sources/Yams/CMakeFiles/Yams.dir/Tag.swift.o Sources/Yams/CMakeFiles/Yams.dir/YamlError.swift.o swift/Yams.swiftmodule 
2022-09-18T05:03:20.5644136Z : && /home/linuxbrew/.linuxbrew/Cellar/swift/5.7/libexec/bin/swiftc -output-file-map Sources/Yams/CMakeFiles/Yams.dir/Release/output-file-map.json -incremental -j 2 -emit-library -o lib/libYams.so -module-name Yams -module-link-name Yams -emit-module -emit-module-path swift/Yams.swiftmodule -emit-dependencies -DSWIFT_PACKAGE -DYams_EXPORTS -module-cache-path "/tmp/swift-20220917-16921-s77h9l/build/swiftpm-linux-x86_64/x86_64-unknown-linux-gnu/yams/module-cache" -O -enable-testing -Xcc -DYAML_DECLARE_STATIC -Xcc -I/tmp/swift-20220917-16921-s77h9l/yams/Sources/CYaml/include -Xcc -F/tmp/swift-20220917-16921-s77h9l/build/foundation-linux-x86_64 -I /tmp/swift-20220917-16921-s77h9l/build/foundation-linux-x86_64/swift /tmp/swift-20220917-16921-s77h9l/yams/Sources/Yams/Constructor.swift /tmp/swift-20220917-16921-s77h9l/yams/Sources/Yams/Decoder.swift /tmp/swift-20220917-16921-s77h9l/yams/Sources/Yams/Emitter.swift /tmp/swift-20220917-16921-s77h9l/yams/Sources/Yams/Encoder.swift /tmp/swift-20220917-16921-s77h9l/yams/Sources/Yams/Mark.swift /tmp/swift-20220917-16921-s77h9l/yams/Sources/Yams/Node.Mapping.swift /tmp/swift-20220917-16921-s77h9l/yams/Sources/Yams/Node.Scalar.swift /tmp/swift-20220917-16921-s77h9l/yams/Sources/Yams/Node.Sequence.swift /tmp/swift-20220917-16921-s77h9l/yams/Sources/Yams/Node.swift /tmp/swift-20220917-16921-s77h9l/yams/Sources/Yams/Parser.swift /tmp/swift-20220917-16921-s77h9l/yams/Sources/Yams/Representer.swift /tmp/swift-20220917-16921-s77h9l/yams/Sources/Yams/Resolver.swift /tmp/swift-20220917-16921-s77h9l/yams/Sources/Yams/String+Yams.swift /tmp/swift-20220917-16921-s77h9l/yams/Sources/Yams/Tag.swift /tmp/swift-20220917-16921-s77h9l/yams/Sources/Yams/YamlError.swift  -Xlinker -soname -Xlinker libYams.so  -L /tmp/swift-20220917-16921-s77h9l/build/swiftpm-linux-x86_64/x86_64-unknown-linux-gnu/yams/lib  -L /tmp/swift-20220917-16921-s77h9l/build/foundation-linux-x86_64/lib   -L /home/linuxbrew/.linuxbrew/lib   -L /usr/lib/gcc/x86_64-linux-gnu/11   -L /home/linuxbrew/.linuxbrew/Cellar/llvm/14.0.6_1/lib -Xlinker -rpath -Xlinker /tmp/swift-20220917-16921-s77h9l/build/foundation-linux-x86_64/lib:  lib/libCYaml.a  -ldispatch  /tmp/swift-20220917-16921-s77h9l/build/foundation-linux-x86_64/lib/libFoundation.so  -lswiftDispatch  -lgcc  -lgcc_s  -lc  -lgcc  -lgcc_s && :
2022-09-18T05:03:20.5644744Z ld.gold: error: lib/libCYaml.a(api.c.o): requires unsupported dynamic reloc 11; recompile with -fPIC
2022-09-18T05:03:20.5645122Z ld.gold: error: lib/libCYaml.a(emitter.c.o): requires unsupported dynamic reloc 11; recompile with -fPIC
2022-09-18T05:03:20.5645558Z ld.gold: error: lib/libCYaml.a(parser.c.o): requires unsupported dynamic reloc 11; recompile with -fPIC
2022-09-18T05:03:20.5645934Z ld.gold: error: lib/libCYaml.a(scanner.c.o): requires unsupported dynamic reloc 11; recompile with -fPIC
2022-09-18T05:03:20.5646284Z ld.gold: error: lib/libCYaml.a(writer.c.o): requires unsupported dynamic reloc 11; recompile with -fPIC
2022-09-18T05:03:20.5646622Z ld.gold: error: lib/libCYaml.a(reader.c.o): requires unsupported dynamic reloc 11; recompile with -fPIC
```

**Comment by Bo98** [comment]
Manually updated Yams to 5.0.1 to fix that issue. Upstream use 5.0.0 but enabling PIC is the only change in 5.0.1.

File: Formula/swift.rb [hunk 1]
```diff
@@ -338,40 +330,30 @@ def install
...
-        # Remove this with Swift 5.7.
-        python_workaround = workspace/"python-workaround"
-        mkdir python_workaround
-        ln_s Utils.safe_popen_read("xcrun", "-find", "python3").chomp, python_workaround/"python"
-        ENV.prepend_path "PATH", python_workaround
       end
       if OS.linux?
         args += %W[
-          --libcxx
+          --libcxx=0
```
Comment (carlocab) [inline Formula/swift.rb:342]: Unused, I presume?

**Review by carlocab — DISMISSED** [review]
Hopefully GitHub doesn't misplace the self-hosted runner while waiting.

File: Formula/swift.rb (hunk 1 above)
Comment (Bo98) [reply Formula/swift.rb:342]: Seems not, or at least not according to https://github.com/apple/swift/commit/e2c5f9bae0e581f8a04658ebf9a1bc116795c26f

**Comment by carlocab** [comment]
Had a peek inside the runner, since `swift` only has a handful of dependents so it shouldn't be taking this long.

It's been stuck testing `vapor` for about 5 hours.

**Comment by carlocab** [comment]
Linux dependents need revision bumps, I think. But that may be better done separately.

**Comment by Bo98** [comment]
Oh oops, swift-format should indeed have been updated alongside.

I don't think anything else should have broke however. Will need to take a look at that.

**Comment by mrstegeman** [comment]
swift-format changes are in #111254

**Comment by bartekpacia** [comment]
Any progress on this? I'd love to see this merged :)

**Comment by SMillerDev** [comment]
You can help fix the build for vapor getting stuck in the test.

**Comment by bartekpacia** [comment]
> You can help fix the build for vapor getting stuck in the test.

I'd love to, if I only had time :). I'm counting on _good open source people_ to deal with this. I didn't mean to be rude or annoying, just wanted to express some interest in getting this merged. Thank you for work!

**Comment by SMillerDev** [comment]
> I'm counting on good open source people to deal with this.

You can be one of good open source people who volunteer their time.

Here's some advice: no pull request ever got merged because random people posted comments expressing interested in getting it merged. Code only gets merged because people put in effort to make it so.

**Comment by Bo98** [comment]
I've had the week off - apologies. Will be looking into this tonight.

**Comment by marcprux** [comment]
Combing over the PR's CI [failure log](https://github.com/Homebrew/homebrew-core/actions/runs/3078070811/jobs/4993456204), it looks like we are still faced with a variety of failures:

1. `brew test --retry --verbose create-api`: this seems like it might be a problem with `create-api` itself.
```log
==> /home/linuxbrew/.linuxbrew/Cellar/create-api/0.1.0/bin/create-api generate /home/linuxbrew/.linuxbrew/Cellar/create-api/0.1.0/share/create-api/test-spec.json --config-option module=TestPackage
  Error: ERROR! The spec is missing or invalid. Expected value to be parsable as Mapping but it was not.
  Generating code for test-spec.json...
```
2. `brew test --retry --verbose publish`: unhelpful error
```log
Error: publish: failed
  An exception occurred within a child process:
    BuildError: Failed executing: /home/linuxbrew/.linuxbrew/Cellar/publish/0.9.0/bin/publish new
  /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:2493:in `block in system'
  /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:2429:in `open'
```
3. `brew test --retry --verbose swift-format`: It appears this was fixed in https://github.com/Homebrew/homebrew-core/pull/111254.
```log
 ==> Testing swift-format
 ==> /home/linuxbrew/.linuxbrew/Cellar/swift-format/0.50600.1/bin/swift-format test.swift
  <unknown>: error: Unable to format test.swift: The loaded '_InternalSwiftSyntaxParser' library is from a toolchain that is not compatible with this version of SwiftSyntax
  
 ==> Testing swift-format (again)
 ==> /home/linuxbrew/.linuxbrew/Cellar/swift-format/0.50600.1/bin/swift-format test.swift
  <unknown>: error: Unable to format test.swift: The loaded '_InternalSwiftSyntaxParser' library is from a toolchain that is not compatible with this version of SwiftSyntax
 Error: swift-format: failed
```
4. `brew test --retry --verbose swiftdraw`: Some kind of XML parsing error
```log
==> /home/linuxbrew/.linuxbrew/Cellar/swiftdraw/0.13.1/bin/swiftdraw /tmp/swiftdraw-test-20220921-137440-1gup5ut/fish.svg --format sfsymbol
Failure: The operation could not be completed. (SwiftDraw.XMLParser.Error error 3.)
```
5. `brew test --retry --verbose vapor`: No output at all. Timeout?
6. `brew test --retry --verbose xclogparser`: Not able to parse a log file. Perhaps the latest version of Xcode made changes to their log format? Probably the package author needs to update their parser.
```log
Error: The file in /tmp/xclogparser-test-20220921-144760-t0jco6/test.xcactivitylog is not a valid SLF log
[567](https://github.com/Homebrew/homebrew-core/actions/runs/3078070811/jobs/4993456204#step:11:578)
```

I'd like to help out with any of these, but I'd like to avoid duplicating tedious debugging. @SMillerDev and @Bo98, can you recommend a good starting task?

**Comment by SMillerDev** [comment]
Trying to reproduce the errors locally is a good start

**Comment by danielnachun** [comment]
I suspect some/all of these just need revision bumps.  Swift 5.7 itself is building fine, it's just some of its dependents that are failing their tests.  Some of them may require changes to support Swift 5.7.

**Comment by marcprux** [comment]
> I suspect some/all of these just need revision bumps. 

What is the policy on this? Is the availability of Swift 5.7 to be held up for everyone due to these projects?

**Comment by danielnachun** [comment]
> What is the policy on this? Is the availability of Swift 5.7 to be held up for everyone due to these projects?

Usually when we do dependent testing we'd like to know why the tests are failing, as they may indicate those packages will need fixes for the newer version of Swift.  That doesn't mean they have to be fixed for the Swift 5.7 to be merged - we just want to understand what may have changed and follow up with fixes soon after.

**Comment by Bo98** [comment]
Revision bumping will indeed fix them. I just had to look into what exactly the issue was and I had very little time this last couple weeks unfortunately.

Basically I think our luck ran out in terms of ABI stability - Swift doesn't guarantee it on Linux (only macOS) and I guess something changes somewhere in a non-obvious way in 5.7.

Basically I'm thinking of moving to versioned `lib` directories (which will break the linkage on version bumps) and/or linking the stdlib statically everywhere, which is something that will happen by default in the future: https://github.com/apple/swift-evolution/blob/main/proposals/0342-static-link-runtime-libraries-by-default-on-supported-platforms.md. I'll push something tomorrow.

**Comment by SMillerDev** [comment]
Seeing how we're trying to get rid of those in every other software I'd hate to have this introduced in swift now.

**Comment by danielnachun** [comment]
> Seeing how we're trying to get rid of those in every other software I'd hate to have this introduced in swift now.

Static linking or versioned library directories?

**Comment by SMillerDev** [comment]
Honestly, both.

**Comment by marcprux** [comment]
> Basically I'm thinking of moving to versioned `lib` directories (which will break the linkage on version bumps) and/or linking the stdlib statically everywhere

Would this have the additional benefit of enabling brew to install swift formula without having to depend on the `swift` formula itself? If so, that would save a ton of CI minutes for workflows that rely on these tools.

**Comment by Bo98** [comment]
> Seeing how we're trying to get rid of those in every other software I'd hate to have this introduced in swift now.

I do agree that it would be nice if things like Go had a small libc-like library. But only if it's ABI stable - I would hate it if we had to revision bump every Go dependent on every patch update and it's a similar story here.

If upstream don't provide ABI stability there's nothing we can really do about it, particularly given static stdlib linking will be a default soon (which AFAIK will still maintain shared linking with dependencies like libxml2, which is nice).

> Would this have the additional benefit of enabling brew to install swift formula without having to depend on the swift formula itself?

It would yes.

**Comment by carlocab** [comment]
Static linking isn't great, but if upstream is going to make it the default, then we may as well go with it.

**Comment by Bo98** [comment]
I've just revision bumped for now. There's an upstream bug that needs fixing before we can use static stdlib linking so I'll revisit the situation in 5.7.1.

I've however added static libraries to the Swift formula with this release since we didn't have them before. `--static-swift-stdlib` will now work, unless you are using FoundationNetworking or FoundationXML.

File: Formula/swift-format.rb [hunk 2]
```diff
@@ -23,14 +23,19 @@ class SwiftFormat < Formula
...
   uses_from_macos "swift"
 
   def install
-    system "swift", "build", "--disable-sandbox", "-c", "release"
+    system "swift", "build", "--disable-sandbox", "-c", "release", "-Xlinker", "-rpath", "-Xlinker", rpath
     bin.install ".build/release/swift-format"
     doc.install "Documentation/Configuration.md"
+
+    if OS.mac?
+      deuniversalize_machos ".build/release/lib_InternalSwiftSyntaxParser.dylib"
```
[the remark this answers is not in the document]
Comment (Bo98) [reply Formula/swift-format.rb:36]: Yeah this was just a direct cherry-pick from the other PR. I will do this.

File: Formula/swift-format.rb [hunk 3]
```diff
@@ -23,14 +23,19 @@ class SwiftFormat < Formula
...
+      lib.install ".build/release/lib_InternalSwiftSyntaxParser.dylib"
```
Comment (carlocab) [inline Formula/swift-format.rb:37]: Also, not sure we want to put this in `lib` where other things could pick it up.

Comment (Bo98) [reply Formula/swift-format.rb:37]: Yes, definitely not.

File: Formula/swift-format.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (carlocab) [reply Formula/swift-format.rb:36]: Or, actually,
```sh
-Xlinker -rpath -Xlinker /Library/Developer/CommandLineTools/usr/lib/swift/macosx
```
is a whole lot simpler.

[the remark this answers is not in the document]
Comment (Bo98) [reply Formula/swift-format.rb:36]: yeah, doing the relative path logic of `rpath(target:` wouldn't be right. Just quickly testing Linux.

[the remark this answers is not in the document]
Comment (carlocab) [reply Formula/swift-format.rb:36]: Opened Homebrew/brew#13978 to make sure we always use absolute paths for these cases.

**Comment by marcprux** [comment]
@Bo98 Can you elaborate on the upstream bug that prevents static linking?

**Comment by Bo98** [comment]
https://github.com/apple/swift/pull/61372

**Comment by marcprux** [comment]
> apple/swift#61372

Bummer. Hopefully @MaxDesiatov can fix it.

**Comment by MaxDesiatov** [comment]
The issue linked here to `apple/swift` repository is now closed. See comments on that issue for more details.

**Comment by Bo98** [comment]
> [apple/swift#61372](https://github.com/apple/swift/pull/61372) is now closed. See comments on that issue for more details.

Thanks for the heads up. Good to know in case we need in the future to send patches to any packages affected. It'll become more of an issue when SE-0342 is implemented so I'll likely wait until then.

**Review by SMillerDev — APPROVED** [review]

**Comment by buttaface** [comment]
> Static linking isn't great, but if upstream is going to make it the default, then we may as well go with it.

Static linking will be the default on linux, but nothing stops [packaging systems like this from using shared libraries instead](https://forums.swift.org/t/pitch-package-manager-statically-link-swift-runtime-libraries-by-default-on-supported-platforms/53900), provided you have some way to deal with the still unstable ABI on non-Darwin platforms.

**Comment by Bo98** [comment]
Yeah, I'll need to think about it. The ABI break from 5.6 -> 5.7 was unfortunate.

Given we're a package manager covering both macOS and Linux, it's just a little annoying dealing with the case where only one OS needs rebuilding.

**Comment by buttaface** [comment]
If you read that pitch thread, it's more an acknowledgement that [most distros do not ship a Swift runtime, stable or not](https://repology.org/project/swift-lang/versions), so static linking allows one to distribute and run Swift binaries on linux easily. As for making the linux ABI stable, somebody will have to drive that and given that [even Apple devs have regrets about ABI decisions they made](https://belkadan.com/blog/2022/10/Swift-in-the-OS/), it may never be worth doing so, just like with Rust and C++.

**Comment by Bo98** [comment]
> As for making the linux ABI stable, somebody will have to drive that and given that [even Apple devs have regrets about ABI decisions they made](https://belkadan.com/blog/2022/10/Swift-in-the-OS/), it may never be worth doing so, just like with Rust and C++.

It's fine to commit to a non-stable ABI. To be clear, I'm not insisting any particular path is taken. The problem is that I don't believe we shouldn't stick on this "maybe" ABI stability.

For Rust, they do commit to non-stability in that if you do choose to link dynamically, you end up linking to something like `libstd-e5c18cc7a05651a4.dylib`, which will change on each version change.

C++ I somewhat dispute. Sure the standard itself doesn't declare an ABI, but implementations do and have been stable for quite some time, which ultimately is what is important. We have libc++1 and libstdc++6.

Swift doesn't have any form of ABI versioning at all. Sure, matching macOS would be very nice, but if that's not possible or desirable then versioning based on what can be guaranteed, e.g. major-minor version, is better than zero versioning.

(Though this discussion is getting better suited for distro chats or somewhere on the forums)

**Pull request closed without merging**